src/Symfony/Component/Process/Process.php

```diff
@@ -1562,4 +1565,28 @@ private function escapeArgument($argument)
 
         return '"'.str_replace(array('"', '^', '%', '!', "\n"), array('""', '"^^"', '"^%"', '"^!"', '!LF!'), $argument).'"';
     }
+
+    private function replacePlaceholder($commandLine, $windows, $env)
+    {
+        $toReplace = $windows ? array('!', '!') : array('$', '');
+        if (!preg_match_all('/\{([a-zA-Z0-9]+?)\}/', $commandLine, $matches)) {
+            return $commandLine;
+        }
+
+        $patterns = $replacements = array();
+        $isInEnv = false;
+        foreach ($matches as $match) {
+            $patterns[] = sprintf('/{%s}/', $match[0]);
+            $replacements[] = sprintf('%s%s%s', $toReplace[0], $match[0], $toReplace[1]);
+            if (isset($env[$match[0]])) {
+                $isInEnv = true;
```

Comment:
Shouldn't you store unused placeholders to display a nice error message like `The placeholders "foo", "bar" are unused`?

Reply:
I can do that, that's a good idea.

Reply:
Any update?

Reply:
updated